_posts/2018-02-15-continuous-deployment-concourse.md

```diff
@@ -0,0 +1,286 @@
+---
+layout: post
+title:  "Continuous deployment to Kubernetes with Concourse CI"
+date:   2018-02-15 11:57:05 -0400
+categories: docker concourse kubernetes
+---
+
+![]({{ "/assets/continuous_deploy_concourse.png" | absolute_url }})
+
+I've been using Concourse CI in conjunction with Kubernetes for a year now and I believe it could provides crucial benefits to every company that is working toward a container shift. 
+
+You have dockerized your application, created Kubernetes templates, why would you stop there and continue using a classic CI. A poorly implemented CI based on plugins, or a hosted solutions that does not integrate well with Docker and where you have no control upon will break your workflow and will force you to do workarounds. On Concourse _everything_ is a container. If you designed your containers well, you won't have to change the way you work between your local environment and Concourse/Kubernetes.
+
+Another massive argument is the scale at which Concourse can operate. As long as you have resources to schedule your containers and volumes, the system will scale. Where I work, we have Concourse workers on an autoscaling group in Google Cloud. At peak traffic we can have dozens of concurrent builds, spinning hundreds of containers and sometimes thousands of volumes.
+
+## Concourse CI notions
+
+- *Resources* are external assets, a Docker image on a registry, a github repository, a slack hook, you name it. They will almost all have checks to trigger jobs, inputs to retrieve and outputs to update/push the resource.
+- *Jobs* contain tasks, these tasks will be executed together in parallel or in sequence on the same node. You will be able to pass inputs and outputs from one task to another.
+- *Tasks* are like functions, with inputs and outputs. Since it's a container you can choose which image it will use to fit your needs.
+
+The *pipeline* contains all the jobs necessary to fulfill a purpose. When a resource gets updated it will trigger a *build* which is the logical execution of the job's tasks.
+
+## Build images
+
+### Docker image resource definition:
+```
+- name: foobar-app-image
+  type: docker-image
+  source:
+    repository: gcr.io/foobar/foobar_app
+    username: _json_key
+    password: {{gcloud_service_key}}
+```
+
+### Build and push the image:
+```
+- put: foobar-app-image
+  params:
+    build: artifact-app/
+    cache: true
+    cache_tag: latest
+    tag: metadata/short-commit-hash
+    tag_as_latest: true
+```
+Note: 
+- the artifact-app/ volume contains a Dockerfile and dependencies to build the image 
+- the `latest` tag will be cached on disk, if we have a lot of builds this can save time!
```

Comment:
- Is this a new sentence? Capitalize the first word and dot the previous line.
- `the latest tag will be cached on disk, if we have a lot of builds this can save time!`
  - This is not a sentence with a relative subordinate, but two sentences. Use a dot.